gpfdist-sink/src/main/java/org/springframework/cloud/stream/module/gpfdist/sink/AbstractGpfdistMessageHandler.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.module.gpfdist.sink;
+
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.integration.handler.AbstractMessageHandler;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHandlingException;
+
+/**
+ * Base implementation of Spring Integration {@code MessageHandler} handling {@code Message}.
+ *
+ * @author Janne Valkealahti
+ */
+public abstract class AbstractGpfdistMessageHandler extends AbstractMessageHandler implements SmartLifecycle {
```

Comment:
No one `MessageHandler` can be `SmartLifecycle`. That is just `Lifecycle` which is propagated from its `Endpoint`:
```
@Override
protected void doStart() {
	this.logComponentSubscriptionEvent(true);
	this.inputChannel.subscribe(this.handler);
	if (this.handler instanceof Lifecycle) {
		((Lifecycle) this.handler).start();
	}
}
```
Otherwise we may end up with the double `start()/stop()` calls.